File: app/src/main/java/com/work/criminalintent/DatePickerFragment.java
package com.work.criminalintent;

import android.app.Activity;
import android.app.Dialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.DatePicker;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;

import java.util.Calendar;
import java.util.Date;
import java.util.GregorianCalendar;

public class DatePickerFragment extends DialogFragment {

    private static final String ARG_DATE = "date";
    public static final String EXTRA_DATE = "com.work.criminalintent.date";

    private DatePicker mDatePicker;

    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
        Date date = (Date) getArguments().getSerializable(ARG_DATE);

        Calendar calendar = Calendar.getInstance();
        calendar.setTime(date);
        int year = calendar.get(Calendar.YEAR);
        int month = calendar.get(Calendar.MONTH);
        int day = calendar.get(Calendar.DAY_OF_MONTH);

        View v = LayoutInflater.from(getActivity())
                .inflate(R.layout.dialog_date, null);

        mDatePicker = v.findViewById(R.id.dialog_date_picker);
        mDatePicker.init(year, month, day, null);
        return new AlertDialog.Builder(getActivity())
                .setTitle(R.string.date_picker_title)
                .setPositiveButton(android.R.string.ok, (dialog, which) -> {
                    int yearPicker = mDatePicker.getYear();
                    int monthPicker = mDatePicker.getMonth();
                    int dayPicker = mDatePicker.getDayOfMonth();
                    Date datePicker = new GregorianCalendar(yearPicker, monthPicker, dayPicker).getTime();
                    sendResult(Activity.RESULT_OK, datePicker);

                })
                .setView(v)
                .create();
    }

    public static DatePickerFragment newInstance(Date date) {
        Bundle args = new Bundle();
        args.putSerializable(ARG_DATE, date);

        DatePickerFragment fragment = new DatePickerFragment();
        fragment.setArguments(args);
        return fragment;
    }

    private void sendResult(int resultCode, Date date) {
        if (getTargetFragment() == null) {
            return;
        }
        Intent intent = new Intent();
        intent.putExtra(EXTRA_DATE, date);
        getTargetFragment().onActivityResult(getTargetRequestCode(), resultCode, intent);
    }
}

File: app/src/main/java/com/work/criminalintent/kt/DatePickerFragment.kt
package com.work.criminalintent.kt

import android.app.Activity
import android.app.AlertDialog
import android.app.Dialog
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.widget.DatePicker
import androidx.fragment.app.DialogFragment
import com.work.criminalintent.R
import java.util.*

const val ARG_DATE = "date"

class DatePickerFragment : DialogFragment() {

    val EXTRA_DATE = "com.work.criminalintent.date";

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val date = arguments?.getSerializable(ARG_DATE) as Date
        val calendar = Calendar.getInstance()
        calendar.time = date
        val year = calendar.get(Calendar.YEAR)
        val month = calendar.get(Calendar.MONTH)
        val day = calendar.get(Calendar.DAY_OF_MONTH)

        val v = LayoutInflater.from(activity)
                .inflate(R.layout.dialog_date, null)

        val mDatePicker = v.findViewById<DatePicker>(R.id.dialog_date_picker)
        mDatePicker.init(year, month, day, null)
        return AlertDialog.Builder(activity)
                .setTitle(R.string.date_picker_title)
                .setPositiveButton(android.R.string.ok) { dialog: DialogInterface, i: Int ->
                    val yearPicker = mDatePicker.year
                    val monthPicker = mDatePicker.month
                    val dayPicker = mDatePicker.dayOfMonth
                    val datePicker = GregorianCalendar(yearPicker, monthPicker, dayPicker).time
                    sendResult(Activity.RESULT_OK, datePicker)
                }
                .setView(v)
                .create()
    }

    private fun sendResult(resultCode: Int, date: Date) {
        if (targetFragment == null) {
            return
        }
        val intent = Intent()
        intent.putExtra(EXTRA_DATE, date)
        targetFragment!!.onActivityResult(targetRequestCode, resultCode, intent)
    }

    companion object {
        fun newInstance(date: Date): DatePickerFragment {
            val args = Bundle()
            args.putSerializable(ARG_DATE, date)
            val fragment = DatePickerFragment()
            fragment.arguments = args
            return fragment
        }

    }
}
